File: Data Structures & Algorithms/find-in-mountain-array/submission-3.py
class Solution:
    def findInMountainArray(self, target: int, mountainArr: 'MountainArray') -> int:
        def search(l, r):
            res = -1
            while l <= r:
                m = (l + r) // 2
                cur = mountainArr.get(m)
                if target <= cur:
                    res = m if target == cur else res
                    r = m-1
                else:
                    l = m + 1
            return res
        
        def searchR(l, r):
            res = -1
            while l <= r:
                m = (l + r) // 2
                cur = mountainArr.get(m)
                if target >= cur:
                    res = m if target == cur else res
                    r = m-1
                else:
                    l = m + 1
            return res

        length = mountainArr.length()
        pivot = (length // 2) - 1 if length % 2 == 0 else (length // 2)
        l, r = 0, length-1
        while l <= r:
            m = (l+r)//2
            left, right, mid = mountainArr.get(m-1), mountainArr.get(m+1), mountainArr.get(m)
            if mid > left and mid > right:
                pivot = m
                break
            elif mid > left and mid < right:
                l = m
            else:
                r = m

        r1 = search(0, pivot)
        r2 = searchR(pivot+1, length-1)

        if r2 == -1:
            return r1
        elif r1 == -1:
            return r2
        else:
            return min(r1, r2)

File: Data Structures & Algorithms/bitwise-and-of-numbers-range/submission-3.rs
impl Solution {
    pub fn range_bitwise_and(mut left: i32, mut right: i32) -> i32 {
        let mut res = 0;

        for i in 0..32 {
            let bit = (left >> i) & 1;
            if bit == 0 {
                continue;
            }

            let diff = (1 << (i+1)) - (left % (1 << (i+1)));

            if right - left < diff {
                res = res | (1 << i);
            }
        }

        res
    }
}
